File: src/main/java/io/klbz/curie/Toggle.java
package io.klbz.curie;

import java.util.Objects;

/**
 * A tiny stateful object that can be either off or on.
 * Useful for detecting and/or testing side effects.
 */
public final class Toggle {
	private boolean isOn;

	private Toggle(boolean isOn){ this.isOn = isOn; }

	public static Toggle on(){ return new Toggle(true); }

	public static Toggle off(){ return new Toggle(false); }

	public void toggle(){ isOn = !isOn; }

	public void turnOn(){ isOn = true; }

	public void turnOff(){ isOn = false; }

	public boolean isOn(){ return isOn; }

	@Override
	public boolean equals(Object o){
		if(this == o) return true;
		if(o == null || getClass() != o.getClass()) return false;
		Toggle toggle = (Toggle) o;
		return isOn == toggle.isOn;
	}

	@Override
	public int hashCode(){ return Objects.hash(isOn); }

	@Override
	public String toString(){ return "Toggle{" + (isOn ? "on" : "off") + '}'; }
}

File: src/main/java/io/klbz/curie/Toggle.kt
package io.klbz.curie

/**
 * A tiny stateful object that can be either off or on.
 * Useful for detecting and/or testing side effects.
 */
class Toggle private constructor(isOn: Boolean) {
    var isOn: Boolean = false
        private set

    init {
        this.isOn = isOn
    }

    fun toggle() {
        isOn = !isOn
    }

    fun turnOn() {
        isOn = true
    }

    fun turnOff() {
        isOn = false
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Toggle

        if (isOn != other.isOn) return false

        return true
    }

    override fun hashCode() = isOn.hashCode()

    override fun toString() = "Toggle{${if (isOn) "on" else "off"}}"

    companion object {
        @JvmStatic
        fun on() = Toggle(true)

        @JvmStatic
        fun off() = Toggle(false)
    }
}
